arangod/RestHandler/RestSystemReportHandler.cpp

```diff
@@ -0,0 +1,145 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2018-2019 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Kaveh Vahedipour
+////////////////////////////////////////////////////////////////////////////////
+
+#include "RestSystemReportHandler.h"
+
+#include "Agency/AgencyFeature.h"
+#include "Agency/Agent.h"
+#include "ApplicationFeatures/ApplicationServer.h"
+#include "Cluster/ServerState.h"
+#include "GeneralServer/ServerSecurityFeature.h"
+#include "Rest/HttpRequest.h"
+#include "Rest/Version.h"
+#include "RestServer/ServerFeature.h"
+#include "Utils/ExecContext.h"
+
+#include <velocypack/Builder.h>
+#include <velocypack/velocypack-aliases.h>
+
+#include <fstream>
+
+using namespace arangodb;
+using namespace arangodb::basics;
+using namespace arangodb::rest;
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief ArangoDB server
+////////////////////////////////////////////////////////////////////////////////
+
+RestSystemReportHandler::RestSystemReportHandler(
+  application_features::ApplicationServer& server, GeneralRequest* request,
+  GeneralResponse* response) : RestBaseHandler(server, request, response) {}
+
+std::string exec(std::string const& cmd) {
+  std::array<char, 128> buffer;
+  std::string result;
+  std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd.c_str(), "r"), pclose);
+  if (!pipe) {
+    throw std::runtime_error("popen() failed!");
+  }
+  while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
+    result += buffer.data();
+  }
+  return result;
+}
+
+bool RestSystemReportHandler::isAdminUser() const {
+  if (!ExecContext::isAuthEnabled()) {
+    return true;
+  } else {
+    return ExecContext::current().isAdminUser();
+  }
+  return false;
+}
+
+std::unordered_map<std::string, std::string> const cmds {
+  {"date", "time date -u \"+%Y-%m-%d %H:%M:%S %Z\" 2>&1"},
+  {"dmesg", "time dmesg 2>&1"},
+  {"df", "time df -h 2>&1"},
+  {"memory", "time cat /proc/meminfo 2>&1"},
+  {"uptime", "time uptime 2>&1"},
+  {"uname", "time uname -a 2>&1"},
+  {"topp", std::string("time top -b -n 1 -H -p ") +
+      std::to_string(Thread::currentProcessId()) + " 2>&1"},
+  {"top", "time top -b -n 1 2>&1"}
```

Comment:
`lib/Basics/process-utils.h` has `ProcessInfo TRI_ProcessInfo(TRI_pid_t pid);` which offers a portable way to get process metrics.